### DISCUSSION TITLE
[DIscussion] Do we need to compatible with IE yet? （3.0 是否还需要支持 IE 问题讨论）

### DISCUSSION THREAD

**Discussion by kuitos** [umijs/qiankun#2822] [General]
Most front-end libraries and frameworks in the community have abandoned support for IE. Is it still necessary to continue to support IE in version 3.0, for example, by providing capabilities like snapshot sandbox?

-------

社区现在大部分前端库、框架都抛弃了对 IE 的支持，3.0 是否还有必要继续兼容 IE，比如提供快照沙箱这样的能力？

**Comment by qiYuei** [comment]
支持的话 `Promise` 、`Generator` 等方法的垫片是用户自行处理吗？感觉没有啥必要支持,毕竟 `2` 已经支持了。`3` 主要面向现代浏览器,有 `IE`方面 需求退回 `2` 即可 😁

**Comment by kuitos** [comment]
现在2需要的polyfill也是要用户自己打的

**Comment by bravepg** [comment]
感觉这个事情还是要考虑投入产出比，不知道市场上用户对 IE 的诉求强不强烈，如果这个时候直接搞上去，没有太多人使用到这条链路就比较尴尬了。

可以先尝试下最小化原则？

**Comment by kuitos** [comment]
要支持的话肯定是有个单独的兼容包，用户引入兼容包后可以支持，关键在于要不要给这个选择

**Comment by wangronglei** [comment]
我选qiankun的愿意主要是兼容ie11，公司要求没办法。就我个人而言，qtmd的ie，就工作而言，我支持提供选择

**Comment by carvinlo** [comment]
个人想法：
1. 使用微前端改造的项目是老项目老系统也有一定占比，否则就只能停留在 v2，缩小了 qiankun 用户群
2. 如果提供ie11兼容的选择成本不高，是否值得支持（比如 patchDocument proxy 改成 defineProperty），功能方面采用渐进增强方式
3. 如果ie11兼容不影响 v3 核心功能，是值得保留

渐进增强方式：并非所有功能都继续支持ie11，比如保留 css 隔离，js 隔离不支持？换一种话说就是，qiankun@3能在ie11上跑起来，冲突比较严重的样式隔离保留，js 冲突问题就让应用在改造 v3 的时候一起整改了，大版本升级本来就要回归测试

**Reply by kuitos** [reply]
如果要兼容 IE 的话，主要工作量来自于：
1. 应用需要在额外打 Stream 相关 API 的 polyfill https://www.npmjs.com/package/web-streams-polyfill
2. 快照沙箱需要保留，同时需要支持 3.0 沙箱里针对 document.head/document.body 的劫持，比较麻烦，可能最简单的方式就是 IE 下先不支持

**Comment by xiaosongxiaosong** [comment]
目前很多政企项目、研究院的项目还是强制要求 ie11 的，一般会在标书上面明确说明，3.0 不支持的话就只能 2.0 一直跑下去

**Reply by xiaosongxiaosong** [reply]
@kuitos 看到一个几年前您介绍的 qiankun 核心价值的文章，个人认为兼容 ie 是支持老旧项目能够继续存活很重要的一个点
https://zhuanlan.zhihu.com/p/95085796

**Comment by Zuojiangtao** [comment]
个人建议保留，市场使用qiankun的项目的团队有相当数量的老旧系统需要渐进式改造。对ie兼容相关的可以拆分为独立插件，如果项目需要自行安装即可。

**Comment by belllee** [comment]
老旧系统可能使用iframe更合适。建议主版本里删除IE的支持，但是可以外挂组件来支持IE
